enterprise/server/util/procstats/procstats.go

```diff
@@ -143,15 +145,19 @@ func pidsInTree(pid int) ([]int, error) {
 		c = append(c, p.Pid())
 		children[ppid] = c
 	}
-	pidsVisited := []int{}
+	pidsVisited := make(map[int]struct{}, len(procs))
 	pidsToExplore := []int{pid}
 	for len(pidsToExplore) > 0 {
 		pid := pidsToExplore[0]
 		pidsToExplore = pidsToExplore[1:]
-		pidsVisited = append(pidsVisited, pid)
+
+		if _, found := pidsVisited[pid]; found {
```

Comment:
The method name refers to a process tree, but we apparently have cycles. Should we rename the method to use terminology that's also correct on Windows? A comment would definitely help.

Reply:
I think https://devblogs.microsoft.com/oldnewthing/20150403-00/?p=44313 is a good read on this topic

> My colleague also pointed out that the thParent­Process­ID is simply a snapshot of the parent process ID at the time the child process was created. Since process IDs can be recycled, once the parent process exits, the process ID is left orphaned, and it may get reassigned to another unrelated process.
> For example, consider this sequence of events:
>
> - Process A creates Process B.
> - Process A terminates, thereby releasing its ID for reuse.
> - Process C is created.
> - Process C reuses Process A’s process ID.

🪟 

Let me add a comment